File: 0131-palindrome-partitioning/0131-palindrome-partitioning.cpp
class Solution {
public:
    bool isPalindrome(string &str,int s,int e){
        while(s<e){
            if(str[s++]!=str[e--]){
                return false;
            }
        }
        return true;
    }
    
    vector<vector<string>> partition(string s) {
        vector<vector<string>> result;
        vector<string>temp;
        solve(0,s.size(),s,temp,result);        
        return result;
    }
    void solve(int idx,int n,string s,vector<string>temp,vector<vector<string>>&result){
        if(idx==n){
            result.push_back(temp);
            return;
        }
        for(int i=idx;i<n;i++){
            if(isPalindrome(s,idx,i)){
                temp.push_back(s.substr(idx,i-idx+1));
                solve(i+1,n,s,temp,result);
                temp.pop_back();
            }
        }
    }
};

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.java
class Solution {
    public List<List<String>> partition(String s) {
        List<List<String>>result= new ArrayList<>();
        solve(0,s.length(),s,new ArrayList<>(),result);
        return result;
    }

    private void solve(int idx, int length, String s, List<String> temp, List<List<String>> result) {
        if(idx==length){
            result.add(new ArrayList<>(temp));
            return;
        }
        for(int i=idx;i<length;i++){
            if(isPalindrome(s,idx,i)){
                temp.add(s.substring(idx,i+1));
                solve(i+1,length,s,temp,result);
                temp.remove(temp.size()-1);
            }
        }
    }

    private boolean isPalindrome(String str, int s, int e) {
        while(s<e){
            if(str.charAt(s)!=str.charAt(e)){
                return false;
            }
            s++;e--;
        }
        return true;
    }
}